### PR TITLE
State t

### PR DESCRIPTION
**Pull request by gabriel128**

### CODE REVIEW

File: src/Course/StateT.hs [hunk 1]
```diff
@@ -1,3 +1,4 @@
+{-# LANGUAGE FlexibleContexts #-}
```
Comment (JulianLeviston) [inline src/Course/StateT.hs:1]: Why flexible contexts?

File: src/Course/StateT.hs [hunk 2]
```diff
@@ -60,102 +52,60 @@ instance Functor f => Functor (StateT s f) where
...
 --
 -- >>> runStateT (state' $ runState $ put 1) 0
 -- ExactlyOne  ((),1)
-state' ::
-  (s -> (a, s))
-  -> State' s a
-state' =
-  error "todo: Course.StateT#state'"
+state' :: (s -> (a, s)) -> State' s a
+state' f =  StateT (pure . f)
```
Comment (JulianLeviston) [inline src/Course/StateT.hs:78]: nice!

File: src/Course/StateT.hs [hunk 3]
```diff
@@ -60,102 +52,60 @@ instance Functor f => Functor (StateT s f) where
...
 -- >>> runState' (state' $ runState $ put 1) 0
 -- ((),1)
-runState' ::
-  State' s a
-  -> s
-  -> (a, s)
-runState' =
-  error "todo: Course.StateT#runState'"
+runState' :: State' s a -> s -> (a, s)
+runState' (StateT fs) = runExactlyOne . fs
```
Comment (JulianLeviston) [inline src/Course/StateT.hs:85]: cool :) you both got this really cleanly. Mine was relatively messy.

File: src/Course/StateT.hs [hunk 4]
```diff
@@ -60,102 +52,60 @@ instance Functor f => Functor (StateT s f) where
...
 
 -- | Run the `State'` seeded with `s` and retrieve the resulting value.
-eval' ::
-  State' s a
-  -> s
-  -> a
-eval' =
-  error "todo: Course.StateT#eval'"
+eval' :: State' s a -> s -> a
+eval' state = runExactlyOne . evalT state
```
Comment (JulianLeviston) [inline src/Course/StateT.hs:101]: Nice... mine is relatively not great.

File: src/Course/StateT.hs [hunk 5]
```diff
@@ -164,25 +114,33 @@ getT =
...
-distinct' ::
-  (Ord a, Num a) =>
-  List a
-  -> List a
-distinct' =
-  error "todo: Course.StateT#distinct'"
+distinct' :: (Ord a, Num a) => List a -> List a
+distinct' xs = eval' (filtering p xs) S.empty
+  where
+    p x = StateT (\s -> ExactlyOne (S.notMember x s, S.insert x s))
```
Comment (JulianLeviston) [inline src/Course/StateT.hs:128]: it's funny... I feel like this pattern you're both using is less clear than using the monadic interface, for some reason. I'm not entirely sure why. It definitely makes things neater and leads to better solutions sometimes, tho, so it's cool :)

File: src/Course/StateT.hs [hunk 6]
```diff
@@ -194,27 +152,24 @@ distinct' =
...
-  (Ord a, Num a) =>
-  List a
-  -> Optional (List a)
-distinctF =
-  error "todo: Course.StateT#distinctF"
+distinctF :: (Ord a, Num a) => List a -> Optional (List a)
+distinctF xs = reverse <$> evalT (filtering p xs) S.empty
+  where
+    p x
+      | x < 100 = StateT (\s -> Full (S.notMember x s, S.insert x s))
```
Comment (JulianLeviston) [inline src/Course/StateT.hs:159]: nice. both of your used guards. I just used stinky old if ;-)

File: src/Course/StateT.hs [hunk 7]
```diff
@@ -194,27 +152,24 @@ distinct' =
...
+data OptionalT f a = OptionalT {runOptionalT :: f (Optional a)}
 
 -- | Implement the `Functor` instance for `OptionalT f` given a Functor f.
 --
 -- >>> runOptionalT $ (+1) <$> OptionalT (Full 1 :. Empty :. Nil)
 -- [Full 2,Empty]
 instance Functor f => Functor (OptionalT f) where
-  (<$>) =
-    error "todo: Course.StateT (<$>)#instance (OptionalT f)"
+  f <$> (OptionalT funct) = OptionalT (((<$>) . (<$>)) f funct)
```
Comment (JulianLeviston) [inline src/Course/StateT.hs:172]: 👍 I quite liked greg's implementation he did for this. I did a similar thing to you.

File: src/Course/StateT.hs [hunk 8]
```diff
@@ -314,21 +275,22 @@ log1 =
...
+-- data OptionalT f a = OptionalT {runOptionalT :: f (Optional a)}
+
+distinctG :: (Integral a, Show a) => List a -> Logger Chars (Optional (List a))
+distinctG xs = filterEmptyness $ runOptionalT (evalT (filtering p (reverse xs)) S.empty)
+  where
+    filterEmptyness (Logger log vals) = Logger (filter (not . isEmpty) log) (reverse <$> vals)
+    logFilter x = if even x then "even number: " ++ show' x else Nil
+    p x
+      | x < 100 = StateT (\s -> OptionalT (log1 (logFilter x) (Full (S.notMember x s, S.insert x s))))
+      | otherwise = StateT (\s -> OptionalT (log1 ("aborting > 100: " ++ show' x) Empty))
```
Comment (JulianLeviston) [inline src/Course/StateT.hs:289]: oh noes... `== 100` would abort...  :)

File: src/Course/StateT.hs (hunk 1 above)
Comment (gabriel128) [reply src/Course/StateT.hs:1]: just stuff that was put by emacs when I was playing with something

File: src/Course/StateT.hs (hunk 5 above)
Comment (gabriel128) [reply src/Course/StateT.hs:128]: I like yours, I want to take the time to understand it, I guess I'll have it now :)

File: src/Course/StateT.hs (hunk 7 above)
Comment (gabriel128) [reply src/Course/StateT.hs:172]: definitely (fmap . fmap) looks nicer, but we don't have that one

File: src/Course/StateT.hs (hunk 8 above)
Comment (gabriel128) [reply src/Course/StateT.hs:289]: oops, forgot that one, I want to refactor this a bit though, this was my first solution

File: src/Course/StateT.hs (hunk 2 above)
Comment (JulianLeviston) [reply src/Course/StateT.hs:78]: So.... if `state' f = StateT (pure . f)`, then `state' = StateT . pure`, yeah? :)

Comment (gabriel128) [reply src/Course/StateT.hs:78]: oh don't look at this PR anymore, I changed most of this, need to finish it though

**Pull request closed without merging**